Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift

```diff
@@ -0,0 +1,139 @@
+    private let operationDispatcher: OperationDispatcher
+    private let productsManager: ProductsManager
+
+    init(receiptFetcher: ReceiptFetcher,
+         introEligibilityCalculator: IntroEligibilityCalculator,
+         backend: Backend,
+         identityManager: IdentityManager,
+         operationDispatcher: OperationDispatcher,
+         productsManager: ProductsManager) {
+        self.receiptFetcher = receiptFetcher
+        self.introEligibilityCalculator = introEligibilityCalculator
+        self.backend = backend
+        self.identityManager = identityManager
+        self.operationDispatcher = operationDispatcher
+        self.productsManager = productsManager
+    }
+
+    func checkEligibility(productIdentifiers: [String],
+                          completionBlock receiveEligibility: @escaping ReceiveIntroEligibilityBlock) {
+        if #available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *) {
+            Task {
+                let eligibility = await sk2CheckEligibility(productIdentifiers)
+                receiveEligibility(eligibility)
+            }
+        } else {
+            sk1CheckEligibility(productIdentifiers, completionBlock: receiveEligibility)
+        }
+    }
+
+    func sk1CheckEligibility(_ productIdentifiers: [String],
+                             completionBlock receiveEligibility: @escaping ReceiveIntroEligibilityBlock) {
+        receiptFetcher.receiptData(refreshPolicy: .onlyIfEmpty) { maybeData in
+            if #available(iOS 12.0, macOS 10.14, tvOS 12.0, watchOS 6.2, *),
+               let data = maybeData {
+                self.sk1CheckEligibility(with: data,
+                                         productIdentifiers: productIdentifiers,
+                                         completionBlock: receiveEligibility)
+            } else {
+                self.backend.getIntroEligibility(appUserID: self.appUserID,
+                                                 receiptData: maybeData ?? Data(),
+                                                 productIdentifiers: productIdentifiers) { result, maybeError in
+                    if let error = maybeError {
+                        Logger.error(Strings.purchase.unable_to_get_intro_eligibility_for_user(error: error))
+                    }
+                    self.operationDispatcher.dispatchOnMainThread {
+                        receiveEligibility(result)
+                    }
+                }
+            }
+        }
+    }
+
+    @available(iOS 12.0, macOS 10.14, tvOS 12.0, watchOS 6.2, *)
+    private func sk1CheckEligibility(with receiptData: Data,
+                                     productIdentifiers: [String],
+                                     completionBlock receiveEligibility: @escaping ReceiveIntroEligibilityBlock) {
+        introEligibilityCalculator
+            .checkEligibility(with: receiptData,
+                              productIdentifiers: Set(productIdentifiers)) { receivedEligibility, maybeError in
```

Comment:
would love to refactor these internal calls to return `Result`, but not for this PR